src/codecs/avif/decoder.rs

```diff
@@ -102,84 +109,369 @@ impl<R: Read> ImageDecoder for AvifDecoder<R> {
+                            }
+                        }
+                    } else {
+                        shape_v_plane();
+                        _bind_v.as_slice()
+                    };
+
+                let worker = match self.picture.pixel_layout() {
+                    PixelLayout::I400 => unreachable!(),
+                    PixelLayout::I420 => yuv420_to_rgba,
+                    PixelLayout::I422 => yuv422_to_rgba,
+                    PixelLayout::I444 => yuv444_to_rgba,
+                };
+
+                let image = YuvPlanarImage::new(
+                    y_plane,
+                    y_plane_stride as usize,
+                    u_plane,
+                    u_plane_stride as usize,
+                    v_plane,
+                    v_plane_stride as usize,
+                    width as usize,
+                    height as usize,
+                );
+
+                let res = worker(
+                    image,
+                    rgba16_buf,
+                    self.picture.bit_depth() as u32,
+                    yuv_range,
+                    color_matrix,
+                );
+
+                if let Err(err) = res {
+                    return Err(ImageError::Unsupported(
+                        UnsupportedError::from_format_and_kind(
+                            ImageFormat::Avif.into(),
+                            UnsupportedErrorKind::GenericFeature(err),
+                        ),
+                    ));
+                }
+            } else {
+                let gray_image = YuvGrayImage::new(
+                    y_plane,
+                    y_plane_stride as usize,
+                    width as usize,
+                    height as usize,
+                );
+                let cr = yuv400_to_rgba(
+                    gray_image,
+                    rgba16_buf,
+                    self.picture.bit_depth() as u32,
+                    yuv_range,
+                    color_matrix,
+                );
+                if let Err(err) = cr {
+                    return Err(ImageError::Unsupported(
+                        UnsupportedError::from_format_and_kind(
+                            ImageFormat::Avif.into(),
+                            UnsupportedErrorKind::GenericFeature(err),
+                        ),
+                    ));
+                }
+            }
+
+            // Squashing alpha plane into a picture
+            if let Some(picture) = self.alpha_picture {
+                if picture.pixel_layout() != PixelLayout::I400 {
+                    return Err(ImageError::Unsupported(
+                        UnsupportedError::from_format_and_kind(
+                            ImageFormat::Avif.into(),
+                            UnsupportedErrorKind::GenericFeature(format!(
+                                "Alpha must be PixelLayout::I400 but was: {:?}",
+                                picture.pixel_layout() // PixelLayout does not implement display
```

Comment:
This one actually was before me.
Yes, `DecodingError` would be best fit here because it normally should not happen, and I think better to consider image as invalid.